src/Symfony/Component/Security/Http/Authenticator/Passport/Badge/UserBadge.php

```diff
@@ -49,20 +52,20 @@ class UserBadge implements BadgeInterface
      * is thrown). If this is not set, the default user provider will be used with
      * $userIdentifier as username.
      */
-    public function __construct(string $userIdentifier, callable $userLoader = null, array $attributes = null)
+    public function __construct(string $userIdentifier, callable $userLoader = null, array $attributes = null, callable $identifierNormalizer = null)
     {
         if (\strlen($userIdentifier) > self::MAX_USERNAME_LENGTH) {
             throw new BadCredentialsException('Username too long.');
         }
-
+        $this->identifierNormalizer = $identifierNormalizer ?? fn ($userIdentifier) => $userIdentifier;
         $this->userIdentifier = $userIdentifier;
         $this->userLoader = $userLoader;
         $this->attributes = $attributes;
     }
 
     public function getUserIdentifier(): string
     {
-        return $this->userIdentifier;
+        return ($this->identifierNormalizer)($this->userIdentifier);
```

Comment:
trying to provide memoization:
```suggestion
        $userIdentifier = $this->userIdentifier ??= ($this->identifierNormalizer)();
        $this->identifierNormalizer = null;
        
        return $userIdentifier;
```

alternative we allow userIdentifier to be a closure and when that's the case, we call it to resolve it to a string

Reply:
Many thanks.
PR up to date.